packages/webdriver/src/node/bidi.ts

```diff
@@ -0,0 +1,136 @@
+import { isIP } from 'node:net'
+import dns from 'node:dns/promises'
+import { type LookupAddress } from 'node:dns'
+
+import logger from '@wdio/logger'
+
+import WebSocket from 'ws'
+
+const log = logger('webdriver')
+const CONNECTION_TIMEOUT = 10000
+
+export async function createBidiConnection(webSocketUrl: string, options?: unknown): Promise<WebSocket | undefined> {
+    const candidateUrls = await listWebsocketCandidateUrls(webSocketUrl)
+    return connectWebsocket(candidateUrls, options)
+}
+
+/**
+ * Resolve hostnames to IPv4 and IPv6 addresses
+ * @returns list of websocket urls to try
+ * @see https://github.com/webdriverio/webdriverio/issues/14039
+ */
+export async function listWebsocketCandidateUrls(webSocketUrl: string): Promise<string[]> {
+    const parsedUrl = new URL(webSocketUrl)
+    const candidateUrls: string[] = [webSocketUrl]
+    if (isIP(parsedUrl.hostname)) {
+        return candidateUrls
+    }
+
+    const candidateIps = await dns.lookup(parsedUrl.hostname, { family:0, all:true })
+    // If the host resolves to a single IP address
+    // then it does not make sense to try additional candidates
+    // as the web socket DNS resolver would do extactly the same
+    if (candidateIps.length > 1) {
+        const hostnameMapper = (result: LookupAddress) => webSocketUrl.replace(parsedUrl.hostname, result.address)
+        candidateUrls.push(...candidateIps.map(hostnameMapper))
+    }
+    return candidateUrls
+}
+
+interface ConnectionResult {
+    ws: WebSocket
+    isConnected: boolean
+    errorMessage?: string
+    index: number
+}
+
+interface ConnectionPromise {
+    index: number
+    promise: Promise<ConnectionResult>
+}
+
+/**
+ * Connect to a websocket
+ * @param candidateUrls - list of websocket urls to try
+ * @returns true if the connection was successful
+ */
+export async function connectWebsocket(candidateUrls: string[], _?: unknown): Promise<WebSocket | undefined> {
+    const websockets: WebSocket[] = candidateUrls.map((candidateUrl) => {
+        log.debug(`Attempt to connect to webSocketUrl ${candidateUrl}`)
+        try {
+            const ws = new WebSocket(candidateUrl)
```

Comment:
Can we send `options` here as second argument in order to be able to send headers when connect to browser using ws? These options are already passed to `connectWebsocket` method